A3-channel-tracing.md

```diff
@@ -0,0 +1,85 @@
+
+Channel tracing is one aspect of the wider design, [channelz](A14-channelz.md).
+
+## Background
+
+Channel connectivity issues are the root cause of a significant portion of user reported gRPC bugs. Channel tracing will be helpful for getting live channel data from a misbehaving program.
+
+## Proposal
+
+To implement channel tracing, implementations will use a dedicated, in-memory _Tracing object_ that is owned by each channel and subchannel. We will add an API that allows library users to retrieve all of the _Tracing object's_ data for the channel as a JSON formatted string. Implementations must expose a toggle for this behavior (for example, C++ will add a new channel argument, GRPC_ARG_CHANNEL_TRACING).
+
+Since the tracing objects may consume large amounts of space, care must be taken to prevent the _Tracing object_ from using too many resources. gRPC implementations should allow limiting the number of _Trace events_ retained to a global maximum, as well as a per channel or subchannel maximum.
+
+Implementations must keep the _Trace data_ for a subchannel until the _Trace event_ signaling that subchannel's destruction is overwritten or garbage collected in the last parent channel to reference it. After that point, implementations are free to use an appropriate garbage collection algorithm to reclaim the memory that the subchannel's _Tracing object_ was holding.
+
+
+## Format of Exported Data
+
+The data will be exported as JSON formatted string. The JSON must conform with the [protobuf-json mapping chart](https://developers.google.com/protocol-buffers/docs/proto3#json). The JSON schema will be as follows:
+
+```
+{
+  "channel_id": string,
+  "channel_data": {
+    // Number of events ever logged in this tracing object. This can differ from
+    // events.size() because events can be overwritten or garbage collected by
+    // implementations.
+    "num_events_logged": number,
+    // Time that this channel was created.
+    "channel_created_timestamp": timestamp string,
+    // List of events that have occurred on this channel.
+    "events": [
+      {
+        // High level description of the event.
+        "description": string,
+        // Status/error associated with this event.
+        // Optional, only present is Status != OK. 
+        "status": string,
+        "time": timestamp string,
+        // can only be one of the states in connectivity_state.h
+        "state": enum string,
+        // uuid of referenced subchannel.
+        // Optional, only present if this event refers to a child object.
+        // and example of a referenced child would be a trace event for a
+        // subchannel being created.
+        "child_channel_id": string
+      },
+    ]
+  },
+  // Optional, only present if this channel has children
+  "child_data": [
```

Comment:
child data should be on the child, no?

Reply:
ref, channel_data, and child_data are all top level to the ChannelTrace proto, so they are also top level on the JSON object.